Index: src/org/eclipse/emf/henshin/interpreter/impl/ChangeImpl.java
===================================================================
--- src/org/eclipse/emf/henshin/interpreter/impl/ChangeImpl.java	(revision 965)
+++ src/org/eclipse/emf/henshin/interpreter/impl/ChangeImpl.java	(revision 1145)
@@ -131,6 +131,12 @@
 		public void applyAndReverse() {
 			// Need to initialize?
 			if (!initialized) {
+			
+			//TODO 20.11.2013
+			//DR: Needs to be revised
+			if (attribute==null) {
+				return;
+			}
 				oldValue = object.eGet(attribute);
 				if ((oldValue==null && newValue==null) ||
 					(oldValue!=null && oldValue.equals(newValue))) {
